### PR TITLE
Settle throttle before running the API GW Tests

### PR DESCRIPTION
**Pull request by mhenke1**
## Description
The API GW Test are able to drive OW into throttling. Though some code existed to 
throttle down these tests these tests were failing constantly on faster machines
with 429 returned by OW.

The test code controlling the throttle was correct in throttling the test execution in isolation,
but was of course not able to overview invocations for the given namspace that do happen 
before tests were executing.

Since this limitations does naturally exist, this change simplifies down the throttling logic
to enhance code maintainability and introduces a wait period before executing the tests to ensure 
that the API GW test do not run into throttling driven by prior invokes in the same namespace.

## Related issue and scope
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: tests/src/test/scala/org/apache/openwhisk/core/cli/test/BaseApiGwTests.scala [hunk 1]
```diff
@@ -56,29 +55,10 @@ abstract class BaseApiGwTests extends TestHelpers with WskTestHelpers with Befor
...
-        s"throttling ${throttlePercent}% of action invocations within last minute = ($numInvocationsToClear) invocations")
-      val throttleTime = 60.seconds.toMillis - (t.toEpochMilli - invocationToClear.toEpochMilli)
-
-      println(s"Waiting ${throttleTime} milliseconds to settle the throttle")
-      Thread.sleep(throttleTime)
-    }
-
-    invocationTimes += Instant.now
+  def throttle(maxInvocationsBeforeThrottle: Int = maxActionsPerMin) = {
+    val throttleTime = (60.seconds.toMillis / maxInvocationsBeforeThrottle * 1.1).toInt
```
Comment (markusthoemmes) [inline tests/src/test/scala/org/apache/openwhisk/core/cli/test/BaseApiGwTests.scala]: Doesn't this lack a relation to `invocationsLast60Seconds` from above?

Comment (mhenke1) [reply tests/src/test/scala/org/apache/openwhisk/core/cli/test/BaseApiGwTests.scala]: @markusthoemmes The original test code worked as expected in isolation. Buit is way to complicated with seeing the limitation that it only controls the test execution of the API GW tests in  and does not prevent all the error situation we have encountered due to prior invocation in the same namespace.

 In decided to dumb it down to throttle each invocation with a shorter fixed period to increase maintainability.

Comment (markusthoemmes) [reply tests/src/test/scala/org/apache/openwhisk/core/cli/test/BaseApiGwTests.scala]: One of the main things I see us losing is time. I agree that functionality-wise this is fine and maybe even better. However, you lose the ability to use all the elapsed time between tests or when not firing any actions (for example when creating them) which the old code took into account.

Comment (mhenke1) [reply tests/src/test/scala/org/apache/openwhisk/core/cli/test/BaseApiGwTests.scala]: @markusthoemmes Sorry for answering late. With the latest commit I reverted back to the original throttling code. I just left the additional pause to calm down the throttle before running the tests.

**Comment by mhenke1** [comment]
@markusthoemmes I decided to make the introduced wait time (which is the only remaining change now) even more explicit while explaining its introduction in a comment. 
Since we get in throttling very often with our newer and faster build environment please have a expedited look.

**Comment by rabbah** [comment]
why do these tests not check the API response which actually provides the reason (and counts) for a 409? the tests could parse that error response and settle the throttle as needed instead of these hacks.

**Comment by rabbah** [comment]
i would not bother and implement a namespace change - it's just another hack.

the 409 response from the controller provides useful information to back off the test and retry correctly:

```
  def tooManyRequests(count: Int, allowed: Int) =
    s"Too many requests in the last minute (count: $count, allowed: $allowed)."

  /** Standard message for too many concurrent activation requests within a time window. */
  def tooManyConcurrentRequests(count: Int, allowed: Int) =
    s"Too many concurrent requests in flight (count: $count, allowed: $allowed)."
```

thread.sleep and using another namespace are just piling on technical debt imo.
